src/R/Core/Impl/AST/Expressions/ExpressionParser.cs

```diff
@@ -449,55 +444,32 @@ enum OperationType {
-            // Indexing or function call is performed on the topmost operand which 
-            // generally should be a variable or a node that evaluates to it.
-            // However, we leave syntax check to separate code.
-
-            IRValueNode operand = this.SafeGetOperand();
-            if (operand == null) {
-                // Oddly, no operand
-                return ParseErrorType.IndentifierExpected;
-            }
-
-            operatorNode.LeftOperand = operand;
-            operatorNode.AppendChild(operand);
-            _operands.Push(operatorNode);
-
-            return ParseErrorType.None;
-        }
-
         private ParseErrorType HandleOperator(ParseContext context, IAstNode parent, out bool isUnary) {
-            ParseErrorType errorType = ParseErrorType.None;
-
-            // If operands stack is empty the operator is unary.
-            // If operator is preceded by another operator, 
-            // it is interpreted as unary.
-
-            TokenOperator currentOperator = new TokenOperator(_operands.Count == 0);
-
+            TokenOperator currentOperator = new TokenOperator(firstInExpression: (_operands.Count == 0));
             currentOperator.Parse(context, null);
             isUnary = currentOperator.IsUnary;
 
+            return HandleOperatorPrecedence(context, currentOperator);
+        }
+
+        private ParseErrorType HandleOperatorPrecedence(ParseContext context, IOperator currentOperator) {
+            ParseErrorType errorType = ParseErrorType.None;
             IOperator lastOperator = _operators.Peek();
-            if (isUnary && lastOperator != null && lastOperator.IsUnary) {
-                // !!!x is treated as !(!(!x)))
-                // Step back and re-parse as expression
-                context.Tokens.Position -= 1;
-                var exp = new Expression(inGroup: false);
-                if (exp.Parse(context, null)) {
-                    _operands.Push(exp);
-                    return ParseErrorType.None;
-                }
-            }
 
-            if (currentOperator.Precedence <= lastOperator.Precedence &&
-                !(currentOperator.OperatorType == lastOperator.OperatorType && currentOperator.Association == Association.Right)) {
+            if (currentOperator.Precedence < lastOperator.Precedence ||
+                (currentOperator.Precedence == lastOperator.Precedence &&
+                 currentOperator.Associativity == Associativity.Left)) {
                 // New operator has lower or equal precedence. We need to make a tree from
                 // the topmost operator and its operand(s). Example: a*b+c. + has lower priority
                 // and a and b should be on the stack along with * on the operator stack.
                 // Repeat until there are no more higher precendece operators on the stack.
 
+                // If operator cas right associativity (exponent, indexer, function call),
```

Comment:
Are indexer and function call actually right-associative? Given that `a[1][2]` is `(a[1])[2]`, and `a(1)(2)` is `(a(1))(2)`.